### PR TITLE
Polynomial ODEs

### LINKED ISSUE #27: Polynomial ODEs

### PR DESCRIPTION
**Pull request by mforets**
Closes #27.

- [x] polynomial system, cont time
- [x] polynomial system, cont time tests
- [x] polynomial system, discr time
- [x] polynomial system, discr time tests

### CODE REVIEW

**Comment by schillic** [comment]
Is `T` a good name for the polynomial type?

**Comment by mforets** [comment]
> Is T a good name for the polynomial type?

How about `PT` or `FT`?

**Comment by schillic** [comment]
Something with `P` makes more sense to me.

**Comment by blegat** [comment]
`PT` or `P` I'd say

**Comment by mforets** [comment]
thanks for commenting, this is ready from my side

File: src/continuous.jl [hunk 1]
```diff
@@ -173,3 +173,46 @@ end
...
+    p::PT
+    statedim::Int
+end
+statedim(s::PolynomialContinuousSystem) = s.statedim
+inputdim(s::PolynomialContinuousSystem) = 0
+
+"""
+    ConstrainedPolynomialContinuousSystem
+
+Continuous-time polynomial system with a constraint set,
```
Comment (schillic) [inline src/continuous.jl]: I find "with state constraints" clearer (we used this description for similar system types).
(same comment for the discrete version)

File: src/continuous.jl [hunk 2]
```diff
@@ -173,3 +173,46 @@ end
...
+- `statedim` -- number of state variables
+"""
+struct ConstrainedPolynomialContinuousSystem{PT, ST} <: AbstractContinuousSystem
+    p::PT
+    statedim::Int
+    X::ST
+end
+statedim(s::ConstrainedPolynomialContinuousSystem) = s.statedim
+stateset(s::ConstrainedPolynomialContinuousSystem) = s.X
+inputdim(s::ConstrainedPolynomialContinuousSystem) = 0
```
Comment (schillic) [inline src/continuous.jl]: Not related, but should we define `inputdim(s::AbstractSystem) = 0` for the abstract supertype, so we do not have to make these trivial definitions for every subtype?

File: test/continuous.jl [hunk 3]
```diff
@@ -78,13 +78,20 @@ end
     @test inputdim(s) == 0
 end
 
-@testset "Initial value problem for a continuous constrained linear algebraic system" begin
-    A = [1. 1; 1 -1]
-    E = [0. 1; 1 0]
```
Comment (schillic) [inline test/continuous.jl]: Why did you remove these tests?

Comment (mforets) [reply test/continuous.jl]: that was unintended 😭

File: src/continuous.jl (hunk 2 above)
Comment (mforets) [reply src/continuous.jl]: it is true that most of the time we have `inputdim` equal to `0`  when there is no input. but i kind of prefer to have to explicitly write the `statedim`, `stateset`, `inputdim` since this makes it clear that this is what you intend for each type...

regarding code duplication what bothers me a bit is Discrete which mostly copy pastes Continuous. and also how different Continuous types are built on top of each other by adding more fields. maybe [Mixers.jl](https://github.com/rafaqz/Mixers.jl) can help..
